### ISSUE TITLE
Unreachability depends on statement ordering

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#18901]
**Bug Report**

`X and sys.platform == "win32"` does not mark the block as unreachable, but `sys.platform == "win32" and X` does. This probably has been reported before but I'm not sure how to phrase this let alone search issues.

**To Reproduce**

```python
import sys

X = True

if X and sys.platform == "win32":
    reveal_type(False)  # N: Revealed type is "Literal[False]?"
```

**Expected Behavior**

I expect the branch to be unreachable, so the `reveal_type` should not show anything

**Your Environment**

Reproduced in mypy play.

- Mypy version used: v1.15
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.12

**Comment by sterliakov** [comment]
Ough. `infer_condition_value` is severely broken...

https://github.com/python/mypy/blob/c7ea0112a3d4cd5db64d11769a6532dec5be5d0a/mypy/reachability.py#L111-L160

Not only it fails to account for your case, but also forgets the negation sometimes, e.g. `if not (sys.platform == 'win32' or sys.platform == 'win32')` will be considered unreachable on Linux, if I'm not too bad at reading this snippet. This may cause a rather funny problem to debug down the road, given that `--warn-unreachable` does not warn on platform-dependent unreachable code, symbols defined there will just be ignored silently along with any errors...

This should be priority-1-normal IMO, but I'm not sure I'm supposed to add priority labels, let's leave this to maintainers?

**Comment by sterliakov** [comment]
Yeah, the negation is indeed lost, [playground](https://mypy-play.net/?mypy=master&python=3.12&flags=strict%2Cwarn-unreachable&gist=9e29a5fd049d2adff9db0a6fcaff04dc):

```python
import sys

if sys.platform == 'linux':
    # sanity check: ensure playground runs on linux
    1 + a  # E

if not sys.platform == 'win32':
    # also good, linux is not win32
    1 + 'a'  # E

if not (sys.platform == 'win32' or sys.platform == 'win32'):
    # well, check it twice and you're doomed
    1 + 'a'
```
